app/questionnaire/questionnaire_manager.py

```diff
@@ -178,134 +44,76 @@ def validate(self, post_data):
             # Not a validation location, so can't be valid
             return False
 
-    def update_node_answers(self, node):
-        answer_dict = {}
-        for answer in self.state.get_answers():
-            answer_value = answer.other if answer.value == 'other' and answer.other else answer.value
-            answer_dict[answer.id] = answer_value
-        node.answers = answer_dict
-
     def validate_all_answers(self):
-        node = self._first
-        valid = True
 
-        while node:
-            self.build_state(node, node.answers)
-            if self.state:
-                self._conditional_display(self.state)
-                is_valid = self.state.schema_item.validate(self.state)
-                if not is_valid:
-                    location = node.item_id
-                    logger.debug("Failed validation with current location %s", location)
-                    # if one of the blocks isn't valid
-                    # then move the current pointer to that block so that the user is redirected to that page
-                    self.go_to_node(location)
-                    break
-
-                logger.debug("Next node is %s", node.item_id)
-
-            node = node.next
-        return valid
-
-    def go_to(self, location):
-        if self._current:
-            logger.debug("Attempting to go to location %s with current location as %s", location, self._current.item_id)
-        if location == 'previous':
-            location = self.navigator.get_previous_location(self.get_answers(), current_location_id=location)
-        elif location == 'summary' and self._current.item_id != 'summary':
-            metadata = get_metadata(current_user)
-            if metadata:
-                logger.warning("User with tx_id %s tried to submit in an invalid state", metadata["tx_id"])
-            raise InvalidLocationException()
-        self.go_to_node(location)
-
-    def is_known_node(self, location):
-        node = self._tail
-        while node:
-            if node.item_id == location:
-                return True
-            node = node.previous
-        return False
+        for location in self.navigator.get_location_path(get_answers(current_user)):
```

Comment:
`get_answers` is called potentially many times, it would be better to call it once and assign it to an `answers` variable. This would make it clear that we don't expect get_answers to return something different on each loop iteration.